File: crates/js_runtime/audio.js
(function(g) {
  'use strict';
  const ops=Deno.core.ops, empty=new Uint8Array();
  const delay=ms=>new Promise(resolve=>setTimeout(resolve,ms));
  class Voice {
    constructor(){this._id=0;this._disposed=false;this._volume=1;this._muted=false;this._loop=false;this._events=new Map();this._timer=null;this._playEpoch=0;}
    addEventListener(type,callback){if(typeof callback==='function'){if(!this._events.has(type))this._events.set(type,new Set());this._events.get(type).add(callback);}}
    removeEventListener(type,callback){this._events.get(type)?.delete(callback);}
    _emit(type,error){const event={type,target:this,error};for(const fn of [...(this._events.get(type)||[])]){try{fn.call(this,event);}catch(e){console.error(e);}}const fn=this['on'+type];if(typeof fn==='function'){try{fn.call(this,event);}catch(e){console.error(e);}}}
    _command(action,value=0){if(this._disposed)throw new Error('Audio disposed');if(!this._id)throw new Error('Audio not loaded');return ops.op_audio_control(this._id,action,value);}
    async _waitReady(id){
      const start=Date.now();
      while(true){
        if(this._disposed||this._id!==id)throw new Error('Audio load cancelled');
        const s=this._command('status');if(s.error)throw new Error(s.error);if(s.ready)return;
        if(Date.now()-start>5000)throw new Error('Audio output unavailable (no active mixer/device)');
        await delay(10);
      }
    }
    _watch(){
      if(this._timer!==null)return;
      const poll=()=>{
        this._timer=null;if(this._disposed||!this._id)return;
        try{const s=this._command('status');if(s.error){this._emit('error',new Error(s.error));return;}this._emit('timeupdate');if(s.ended){this._emit('ended');return;}if(!s.paused)this._timer=setTimeout(poll,50);}
        catch(e){this._emit('error',e);}
      };
      this._timer=setTimeout(poll,50);
    }
    pause(){this._playEpoch++;if(this._id)this._command('pause');if(this._timer!==null){clearTimeout(this._timer);this._timer=null;}this._emit('pause');}
    get paused(){return !this._id||this._command('status').paused;}
    get ended(){return !!this._id&&this._command('status').ended;}
    get currentTime(){return this._id?this._command('status').currentTime:0;}
    get duration(){return this._id?(this._command('status').duration??Infinity):NaN;}
    get volume(){return this._volume;}
    set volume(value){value=Number(value);if(!Number.isFinite(value)||value<0||value>1)throw new RangeError('Volume must be 0..1');this._volume=value;if(this._id)this._command('volume',this._muted?0:value);}
    get muted(){return this._muted;}
    set muted(value){this._muted=!!value;if(this._id)this._command('volume',this._muted?0:this._volume);}
    _release(){if(this._timer!==null)clearTimeout(this._timer);this._timer=null;if(this._id){ops.op_audio_control(this._id,'dispose',0);this._id=0;}}
    dispose(){if(this._disposed)return;this._playEpoch++;this._release();this._disposed=true;this._events.clear();}
  }
  class Audio extends Voice {
    constructor(src=''){super();this._src=src;this._generation=0;this._loading=null;}
    get src(){return this._src;}
    set src(value){if(this._disposed)throw new Error('Audio disposed');this._generation++;this._playEpoch++;this._release();this._loading=null;this._src=value;}
    get loop(){return this._loop;}
    set loop(value){this._loop=!!value;if(this._id)this._command('loop',this._loop?1:0);}
    get currentTime(){return super.currentTime;}
    set currentTime(value){this._command('seek',Number(value));}
    load(){
      if(this._disposed)return Promise.reject(new Error('Audio disposed'));
      if(this._loading)return this._loading;
      if(this._id)return Promise.resolve();
      const generation=this._generation;
      this._loading=(async()=>{
        let bytes;
        if(this._src instanceof Blob)bytes=new Uint8Array(await this._src.arrayBuffer());
        else if(this._src instanceof ArrayBuffer||ArrayBuffer.isView(this._src))bytes=__lunaBinary.view(this._src);
        else {if(!this._src)throw new TypeError('Audio source is empty');const response=await fetch(String(this._src));if(!response.ok)throw new Error('Audio HTTP '+response.status);bytes=new Uint8Array(await response.arrayBuffer());}
        if(this._disposed||generation!==this._generation)throw new Error('Audio load cancelled');
        const id=ops.op_audio_create('clip',bytes,0,0,0);this._id=id;
        this._command('volume',this._muted?0:this._volume);this._command('loop',this._loop?1:0);
        await this._waitReady(id);this._emit('loadeddata');
      })().catch(e=>{if(generation===this._generation&&!this._disposed){this._release();this._emit('error',e);}throw e;}).finally(()=>{if(generation===this._generation)this._loading=null;});
      this._loading.catch(()=>{});return this._loading;
    }
    async play(){const epoch=++this._playEpoch;await this.load();if(epoch!==this._playEpoch)throw new Error('Audio play cancelled');this._command('play');this._emit('play');this._watch();}
    stop(){this.pause();if(this._id)this.currentTime=0;}
  }
  class AudioStream extends Voice {
    constructor({sampleRate=48000,channels=2,bufferSeconds=2}={}){
      super();
      if(!Number.isInteger(sampleRate)||!Number.isInteger(channels))throw new TypeError('Invalid PCM format');
      this._id=ops.op_audio_create('stream',empty,sampleRate,channels,bufferSeconds);
      Object.defineProperties(this,{sampleRate:{value:sampleRate},channels:{value:channels}});
      this._ready=this._waitReady(this._id).catch(e=>{this._release();this._emit('error',e);throw e;});this._ready.catch(()=>{});
    }
    get ready(){return this._ready;}
    async play(){const epoch=++this._playEpoch;await this.ready;if(epoch!==this._playEpoch)throw new Error('Audio play cancelled');this._command('play');this._emit('play');this._watch();}
    get bufferedFrames(){return this._command('status').bufferedFrames;}
    get writableFrames(){return this._command('status').writableFrames;}
    write(samples){
      if(this._disposed)throw new Error('Audio disposed');
      if(!(samples instanceof Float32Array)||samples.length%this.channels)throw new TypeError('Expected interleaved Float32Array with complete frames');
      return ops.op_audio_write(this._id,__lunaBinary.view(samples));
    }
    async writeAll(samples){
      if(this._writing)throw new Error('Only one writeAll may be pending');this._writing=true;
      try{let frame=0;while(frame*this.channels<samples.length){const n=this.write(samples.subarray(frame*this.channels));frame+=n;if(!n)await delay(10);}return frame;}finally{this._writing=false;}
    }
    end(){this._command('end');}
  }
  g.Audio=Audio;g.AudioStream=AudioStream;
})(globalThis);

File: crates/js_runtime/src/audio.rs
//! Isolate-owned audio. Decoding runs on the JS worker, never on Bevy's render thread.
use deno_core::{op2, OpState};
use rodio::Source;
use serde::Serialize;
use std::{
    collections::{HashMap, VecDeque},
    io::Cursor,
    sync::{Arc, Mutex},
    time::Duration,
};
const MAX_BYTES: usize = 64 * 1024 * 1024;
const MAX_VOICES: usize = 16;

pub enum Samples {
    Clip(Vec<f32>),
    Stream {
        queue: VecDeque<f32>,
        capacity: usize,
        eof: bool,
    },
}
pub struct Playback {
    pub samples: Samples,
    pub channels: u16,
    pub rate: u32,
    pub cursor: usize,
    pub playing: bool,
    pub looping: bool,
    pub volume: f32,
    pub disposed: bool,
    pub ready: bool,
    pub allowed: bool,
    pub error: Option<String>,
    pub ended: bool,
}
pub type SharedPlayback = Arc<Mutex<Playback>>;
#[derive(Default)]
pub struct AudioQueue {
    next: u32,
    entries: HashMap<u32, SharedPlayback>,
    pending: Vec<SharedPlayback>,
    sizes: HashMap<u32, usize>,
}
impl Drop for AudioQueue {
    fn drop(&mut self) {
        for p in self.entries.values().chain(self.pending.iter()) {
            if let Ok(mut p) = p.lock() {
                p.disposed = true;
                p.samples = Samples::Clip(Vec::new());
            }
        }
    }
}
impl AudioQueue {
    pub fn drain(&mut self) -> Vec<SharedPlayback> {
        std::mem::take(&mut self.pending)
    }
    fn get(&self, id: u32) -> Result<SharedPlayback, anyhow::Error> {
        self.entries
            .get(&id)
            .cloned()
            .ok_or_else(|| anyhow::anyhow!("Unknown or disposed audio in this isolate"))
    }
}
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Status {
    ready: bool,
    paused: bool,
    ended: bool,
    current_time: f64,
    duration: Option<f64>,
    buffered_frames: usize,
    writable_frames: usize,
    error: Option<String>,
}
impl Playback {
    fn status(&self) -> Status {
        let (duration, buffered, free) = match &self.samples {
            Samples::Clip(v) => (
                Some(v.len() as f64 / self.channels as f64 / self.rate as f64),
                0,
                0,
            ),
            Samples::Stream {
                queue, capacity, ..
            } => (
                None,
                queue.len() / self.channels as usize,
                (capacity - queue.len()) / self.channels as usize,
            ),
        };
        Status {
            ready: self.ready,
            paused: !self.playing,
            ended: self.ended,
            current_time: self.cursor as f64 / self.channels as f64 / self.rate as f64,
            duration,
            buffered_frames: buffered,
            writable_frames: free,
            error: self.error.clone(),
        }
    }
}
#[op2(fast)]
#[smi]
pub fn op_audio_create(
    state: &mut OpState,
    #[string] kind: String,
    #[buffer] bytes: &[u8],
    sample_rate: u32,
    channels: u32,
    seconds: f64,
) -> Result<u32, anyhow::Error> {
    let q = state.borrow_mut::<AudioQueue>();
    anyhow::ensure!(
        q.entries.len() < MAX_VOICES && q.pending.len() < MAX_VOICES * 2,
        "Audio voice/creation queue limit reached"
    );
    q.pending.retain(|p| !p.lock().unwrap().disposed);
    let available = MAX_BYTES - q.sizes.values().sum::<usize>();
    let (samples, channels, rate, size) = if kind == "clip" {
        anyhow::ensure!(
            bytes.len() <= 8 * 1024 * 1024,
            "Encoded audio exceeds 8 MiB"
        );
        let decoder = rodio::Decoder::new(Cursor::new(bytes.to_vec()))?;
        let channels = decoder.channels();
        let rate = decoder.sample_rate();
        anyhow::ensure!(
            (1..=2).contains(&channels) && (8000..=192000).contains(&rate),
            "Audio must be mono/stereo, 8..192 kHz"
        );
        let mut data: Vec<f32> = decoder
            .convert_samples::<f32>()
            .take(available / 4 + 1)
            .collect();
        anyhow::ensure!(
            !data.is_empty() && data.len() * 4 <= available && data.len() % channels as usize == 0,
            "Decoded audio exceeds quota or has no complete frames"
        );
        anyhow::ensure!(
            data.iter().all(|v| v.is_finite()),
            "Decoded audio contains invalid samples"
        );
        for v in &mut data {
            *v = v.clamp(-1., 1.);
        }
        let size = data.len() * 4;
        (Samples::Clip(data), channels, rate, size)
    } else {
        anyhow::ensure!(kind == "stream", "Unknown audio source kind");
        anyhow::ensure!(
            (1..=2).contains(&channels) && (8000..=192000).contains(&sample_rate),
            "PCM must be mono/stereo, 8..192 kHz"
        );
        anyhow::ensure!(
            seconds.is_finite() && (0.1..=10.0).contains(&seconds),
            "Buffer duration must be 0.1..10 seconds"
        );
        let capacity = (seconds * sample_rate as f64).ceil() as usize * channels as usize;
        anyhow::ensure!(capacity * 4 <= available, "Audio memory quota exceeded");
        (
            Samples::Stream {
                queue: VecDeque::with_capacity(capacity),
                capacity,
                eof: false,
            },
            channels as u16,
            sample_rate,
            capacity * 4,
        )
    };
    let p = Arc::new(Mutex::new(Playback {
        samples,
        channels,
        rate,
        cursor: 0,
        playing: false,
        looping: false,
        volume: 1.,
        disposed: false,
        ready: false,
        allowed: false,
        error: None,
        ended: false,
    }));
    q.next = q
        .next
        .checked_add(1)
        .ok_or_else(|| anyhow::anyhow!("Audio ID exhausted"))?;
    let id = q.next;
    q.sizes.insert(id, size);
    q.entries.insert(id, p.clone());
    q.pending.push(p);
    Ok(id)
}
#[op2]
#[serde]
pub fn op_audio_control(
    state: &mut OpState,
    id: u32,
    #[string] action: String,
    value: f64,
) -> Result<Status, anyhow::Error> {
    let q = state.borrow_mut::<AudioQueue>();
    let shared = q.get(id)?;
    let mut p = shared.lock().unwrap();
    match action.as_str() {
        "status" => {}
        "dispose" => {
            p.disposed = true;
            p.playing = false;
            p.samples = Samples::Clip(Vec::new());
            q.entries.remove(&id);
            q.sizes.remove(&id);
        }
        "play" => {
            anyhow::ensure!(!p.disposed, "Audio disposed");
            if let Some(e) = &p.error {
                anyhow::bail!(e.clone());
            }
            if p.ended {
                match &p.samples {
                    Samples::Clip(_) => p.cursor = 0,
                    _ => anyhow::bail!("An ended PCM stream cannot restart"),
                }
            }
            p.ended = false;
            p.playing = true;
        }
        "pause" => p.playing = false,
        "volume" => {
            anyhow::ensure!(
                value.is_finite() && (0.0..=1.0).contains(&value),
                "Volume must be 0..1"
            );
            p.volume = value as f32;
        }
        "loop" => {
            anyhow::ensure!(
                matches!(p.samples, Samples::Clip(_)),
                "Live streams cannot loop"
            );
            p.looping = value != 0.;
        }
        "seek" => {
            anyhow::ensure!(value.is_finite() && value >= 0., "Invalid playback time");
            let Samples::Clip(v) = &p.samples else {
                anyhow::bail!("Live streams cannot seek")
            };
            p.cursor = ((value * p.rate as f64).floor() as usize)
                .saturating_mul(p.channels as usize)
                .min(v.len());
            p.ended = false;
        }
        "end" => {
            let Samples::Stream { eof, .. } = &mut p.samples else {
                anyhow::bail!("Not a PCM stream")
            };
            *eof = true;
        }
        _ => anyhow::bail!("Unknown audio operation"),
    }
    Ok(p.status())
}
#[op2(fast)]
#[smi]
pub fn op_audio_write(
    state: &mut OpState,
    id: u32,
    #[buffer] bytes: &[u8],
) -> Result<u32, anyhow::Error> {
    let shared = state.borrow::<AudioQueue>().get(id)?;
    let mut p = shared.lock().unwrap();
    anyhow::ensure!(!p.disposed, "Audio disposed");
    if let Some(e) = &p.error {
        anyhow::bail!(e.clone());
    }
    let channels = p.channels as usize;
    anyhow::ensure!(
        bytes.len() % (channels * 4) == 0,
        "PCM must contain complete interleaved Float32 frames"
    );
    let Samples::Stream {
        queue,
        capacity,
        eof,
    } = &mut p.samples
    else {
        anyhow::bail!("Not a PCM stream")
    };
    anyhow::ensure!(!*eof, "PCM stream already ended");
    let n = (bytes.len() / 4).min(*capacity - queue.len());
    let values: Vec<f32> = bytes[..n * 4]
        .chunks_exact(4)
        .map(|b| f32::from_le_bytes(b.try_into().unwrap()))
        .collect();
    anyhow::ensure!(
        values.iter().all(|v| v.is_finite() && v.abs() <= 1.),
        "PCM samples must be finite and between -1 and 1"
    );
    queue.extend(values);
    Ok((n / channels) as u32)
}

/// Mixer reads small blocks, with no allocation or network work in the callback.
/// Temporary PCM starvation produces silence; EOF is explicit, never inferred.
pub struct AudioDecoder {
    shared: SharedPlayback,
    block: [f32; 256],
    at: usize,
    len: usize,
    channels: u16,
    rate: u32,
}
impl AudioDecoder {
    pub fn new(shared: SharedPlayback) -> Self {
        let (channels, rate) = {
            let mut p = shared.lock().unwrap();
            p.ready = true;
            (p.channels, p.rate)
        };
        Self {
            shared,
            block: [0.; 256],
            at: 0,
            len: 0,
            channels,
            rate,
        }
    }
}
impl Iterator for AudioDecoder {
    type Item = f32;
    fn next(&mut self) -> Option<f32> {
        if self.at == self.len {
            // Never wait for the JS producer on the device's audio thread.
            match self.shared.try_lock() {
                Ok(mut p) => {
                    if p.disposed {
                        return None;
                    }
                    self.block.fill(0.);
                    self.len = 256;
                    if p.playing && p.allowed {
                        for out in &mut self.block {
                            let cursor = p.cursor;
                            let looping = p.looping;
                            let next = match &mut p.samples {
                                Samples::Clip(v) => {
                                    if cursor < v.len() {
                                        Some(v[cursor])
                                    } else if looping {
                                        Some(v[cursor % v.len()])
                                    } else {
                                        None
                                    }
                                }
                                Samples::Stream { queue, .. } => queue.pop_front(),
                            };
                            if let Some(v) = next {
                                *out = v * p.volume;
                                p.cursor += 1;
                                if p.looping {
                                    if let Samples::Clip(v) = &p.samples {
                                        p.cursor %= v.len();
                                    }
                                }
                            } else {
                                let eof = match &p.samples {
                                    Samples::Clip(_) => true,
                                    Samples::Stream { eof, .. } => *eof,
                                };
                                if eof {
                                    p.ended = true;
                                    p.playing = false;
                                }
                                break;
                            }
                        }
                    }
                }
                Err(std::sync::TryLockError::WouldBlock) => {
                    self.block.fill(0.);
                    self.len = 256;
                }
                Err(_) => return None,
            }
            self.at = 0;
        }
        let value = self.block[self.at];
        self.at += 1;
        Some(value)
    }
}
impl Source for AudioDecoder {
    fn current_frame_len(&self) -> Option<usize> {
        None
    }
    fn channels(&self) -> u16 {
        self.channels
    }
    fn sample_rate(&self) -> u32 {
        self.rate
    }
    fn total_duration(&self) -> Option<Duration> {
        None
    }
}
